src/mito2/src/schedule/priority.rs

```diff
@@ -0,0 +1,447 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Add;
+use std::pin::Pin;
+use std::time::{Duration, Instant};
+
+use async_stream::stream;
+use futures::Stream;
+use tokio::select;
+use tokio::time::Sleep;
+
+use crate::error;
+
+/// Creates an unbounded channel.
+pub fn unbounded<T>() -> (Sender<T>, Receiver<T>) {
+    let (tx_low, rx_low) = async_channel::unbounded();
+    let (tx_high, rx_high) = async_channel::unbounded();
+    (Sender { tx_low, tx_high }, Receiver { rx_low, rx_high })
+}
+
+#[derive(Clone)]
+pub struct Sender<T> {
+    tx_low: async_channel::Sender<(Option<Instant>, T)>,
+    tx_high: async_channel::Sender<T>,
+}
+
+impl<T> Sender<T> {
+    /// Tries to send a high priority item, returns `Err` if the channel is full or closed.
+    pub fn try_send_high(&self, item: T) -> error::Result<()> {
+        self.tx_high.try_send(item).map_err(map_send_error)
+    }
+
+    /// Tries to send a low priority item with optional deadline, returns `Err` if the channel is full or closed.
+    pub fn try_send_low(&self, item: T, deadline: Option<Duration>) -> error::Result<()> {
+        self.tx_low
+            .try_send((deadline.map(|d| Instant::now().add(d)), item))
+            .map_err(map_send_error)
+    }
+
+    /// Tries to send a high priority item, returns `Err` if the channel is full or closed.
+    #[cfg(test)]
+    pub async fn send_high(&self, item: T) -> error::Result<()> {
+        self.tx_high.send(item).await.map_err(map_send_error)
+    }
+
+    /// Tries to send a low priority item with optional deadline, returns `Err` if the channel is full or closed.
+    #[cfg(test)]
+    pub async fn send_low(&self, item: T, deadline: Option<Duration>) -> error::Result<()> {
+        self.tx_low
+            .send((deadline.map(|d| Instant::now().add(d)), item))
+            .await
+            .map_err(map_send_error)
+    }
+}
+
+fn map_send_error<E>(e: E) -> error::Error
+where
+    E: std::fmt::Debug,
+{
+    error::SendToChannelSnafu {
+        err_msg: format!("{:?}", e),
+    }
+    .build()
+}
+
+/// Channel receiver
+pub struct Receiver<T> {
+    rx_low: async_channel::Receiver<(Option<Instant>, T)>,
+    rx_high: async_channel::Receiver<T>,
+}
+
+impl<T> Clone for Receiver<T> {
+    fn clone(&self) -> Self {
+        Self {
+            rx_high: self.rx_high.clone(),
+            rx_low: self.rx_low.clone(),
+        }
+    }
+}
+
+impl<T> Receiver<T> {
```

Comment:
I wonder if separating workers for low and high priorities like rocksdb is more robust
https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide#parallelism-options

Each channel services a worker pool. Then if the number of low priority workers is less than the number of global runtime workers, we can ensure compaction never blocks flush. (We may need to limit the concurrency of a single compaction task, now it spawns many sub tasks)

Reply:
> I wonder if separating workers for low and high priorities like rocksdb is more robust https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide#parallelism-options
> 
> Each channel services a worker pool. Then if the number of low priority workers is less than the number of global runtime workers, we can ensure compaction never blocks flush. (We may need to limit the concurrency of a single compaction task, now it spawns many sub tasks)

I tried not to touch the scheduler itself in this PR, but tbh it's a bit confusing, especially in that flush jobs are executed in place while compaction tasks are spawned to separate runtime but the scheduler still has to wait for it to complete. I did some benchmark but not yet decided whether to refactor the scheduler so that we can  control the parallelism a in better manner.

Reply:
We should avoid spawning tasks inside the compaction job. We might create a job for each output instead of processing jobs in parallel inside the scheduler.

> compaction tasks are spawned to separate runtime but the scheduler still has to wait for it to complete.

Yes, it also blocks the scheduler. Maybe the scheduler for compaction should run in the compact runtime.

> not yet decided whether to refactor the scheduler so that we can control the parallelism a in better manner.

We may refactor the local compactor first. What's more, we may need a different concurrency control mechanism if we use a remote compactor. A trivial approach is providing a larger max compaction jobs limit if we use the remote compactor.

Maybe we can implement the scheduler in a simple style: It just has a semaphore and always spawns the job and acquires the semaphore.